sdk/core/core/src/az_span.c

```diff
@@ -153,6 +154,30 @@ AZ_NODISCARD az_result az_span_atou32(az_span span, uint32_t* out_number)
   return AZ_OK;
 }
 
+AZ_NODISCARD az_result az_span_atod(az_span span, az_span* conversion_buffer, double* out_number)
+{
+  AZ_PRECONDITION_VALID_SPAN(span, 1, false);
+  AZ_PRECONDITION_NOT_NULL(out_number);
+
+  if (conversion_buffer == NULL)
+  {
+    // This means there is already an invalid character after the last digit to convert
+    *out_number = atof((char*)az_span_ptr(span));
+    return AZ_OK;
```

Comment:
why would we return an error?  that would be right.
So, we are just offering an easy way to use atof() directly with az_span, we should not make it more complicated by validating or providing errors that customers needs to learn

Reply:
What is the contract; not the implementation.
Do we want our az_span_atod function to always return 0 even if error?
I think this would be OK (since it's idiomatic C) and then we'd return double and not have an out parameter.

If we want the contract to return error (not 0) for "Hello", then we have to do more than JUST call atod; we'd hae to check for valid digits  and period first and either return error or the result of atod.

Reply:
We need the az_result because we ask for an extra buffer to copy before calling atof().
If the buffer has not enogh size for the copy we return ERROR.

That's the only thing we consider as error

Reply:
That isn't true for the other APIs, like atou64, where we return unexpected char, why is it true for atod?
https://github.com/Azure/azure-sdk-for-c/blob/bed46d6d13f0783168e101b17eae738035cc5600/sdk/core/core/src/az_span.c#L113

Returning 0 is certainly odd to me. How as the caller am i supposed to distinguish between a valid scenario and a dev bug?

az_span invalid = "hello";
az_span valid = "0";

Both of these will return the 0.0 double.

Reply:
Is it feasible to use strtod?
https://stackoverflow.com/questions/30481587/alternative-to-atof-function-to-differenciate-input-0-with-failure

Reply:
Why do we need atod? I have this conversation buffer and copy; it's memory intensive and CPU intensive and an interior programming model. It's all lose. How badly do we need this function? Can it be internal if we have to have it?

If we must have it, then I think we need our own implementation.

Reply:
I don't know if there currently is a need for this, but it is just the mirror of the dtoa function.
This goes back to the issue which discusses being able to roundtrip things and parse/format the 5 number formats. I guess the principle question is, if we support formatting something, shouldn't we also support parsing it and vice versa?
https://github.com/Azure/azure-sdk-for-c/issues/598

@vhvb1989 - please correct me if I am mistaken or if you already know of a use case.

> I have this conversation buffer and copy; it's memory intensive and CPU intensive and an interior programming model

Agreed, but presumably (and hopefully) most callers would just pass NULL for that and not do any data copies.

> If we must have it, then I think we need our own implementation.

That is a tricky implementation but might be worth the trade off.

Reply:
Yes, just like Ahson mentioned.  It is not used anywhere. We want it to mirror our az_span_dtoa.
However, our current az_span_dtoa() can't work with decimals anyway.

I am more than happy to kill this function!  :)  I consider this one and others as a super set of advanced functionality we are trying to offer to customers in advance and before customer really ask for it. So, yes, I would be happy to wait for actually customer asking for this and just leave the issue open as reference.

Reply:
OK, kill it for now.

Reply:
I discussed this a bit with Cris, and I think customers will need double support for building/parsing JSON (for things like temperature values). It is also possible that PnP will need support for this in the near future.